compiler/src/lnntoamm/Statement.ts

```diff
@@ -0,0 +1,785 @@
+import { LPNode, NamedAnd, NamedOr } from "../lp";
+import Output from "./Amm";
+import Event from './Event'
+import Fn from "./Fn";
+import opcodes from "./opcodes";
+import Scope from "./Scope";
+import Type, { Builtin, FunctionType } from "./Types";
+import { genName, TODO } from "./util";
+
+export class MetaData {
+  scope: Scope
+  variables: { [name: string]: Dec }
+  retConstraints: Type[]
+
+  constructor(
+    scope: Scope,
+    variables: { [name: string]: Dec } = null,
+    retConstraints: Type[] = null,
+  ) {
+    this.scope = scope;
+    this.variables = variables !== null ? variables : {};
+    this.retConstraints = retConstraints !== null ? retConstraints : [];
+  }
+
+  var(name: string): Dec {
+    if (this.variables[name] == null) {
+      return null;
+    }
+    return this.variables[name];
+  }
+
+  define(dec: Dec) {
+    if (this.var(dec.name) !== null) {
+      throw new Error(`Can't redefine value ${dec.name}`);
+    }
+    this.variables[dec.name] = dec;
+  }
+}
+
+export default abstract class Stmt {
+  ast: LPNode
+
+  constructor(
+    ast: LPNode,
+  ) {
+    this.ast = ast;
+  }
+
+  // interface fns
+  abstract exprTy(): Type;
+  abstract inline(amm: Output): void;
+
+  // =================
+  // factory functions
+  static fromAst(ast: LPNode, metadata: MetaData): Stmt[] {
+    let stmts = [];
+    if (ast.has('assignables')) {
+      stmts.push(...Stmt.fromAssignables(ast.get('assignables'), metadata));
+    } else if (ast.has('assignments')) {
+      stmts.push(...Assign.fromAssignmentsAst(ast.get('assignments'), metadata));
+    } else if (ast.has('conditionals')) {
+      stmts.push(...Cond.fromConditionalsAst(ast.get('conditionals'), metadata));
+    } else if (ast.has('declarations')) {
+      stmts.push(...Dec.fromAst(ast.get('declarations'), metadata));
+    } else if (ast.has('emits')) {
+      stmts.push(...Emit.fromAst(ast.get('emits'), metadata));
+    } else if (ast.has('exits')) {
+      stmts.push(...Exit.fromAst(ast.get('exits'), metadata));
+    } else {
+      throw new Error(`unrecognized statement ast: ${ast}`);
+    }
+    return stmts;
+  }
+
+  static fromAssignables(ast: LPNode, metadata: MetaData): Stmt[] {
+    let stmts = [];
+
+    let asts: LPNode[] = ast.getAll().map(a => a.get('withoperators'));
+    if (asts.length > 1) TODO('operators');
+    asts = asts.length === 1 ? asts.pop().get('baseassignablelist').getAll().map(a => a.get('baseassignable')) : [];
```

Comment:
Nit: may want to check if the only value is a `baseassignablelist` and throw an error if the only value is an operator?